### PR TITLE
vkd3d: Add new interfaces: ID3D12DeviceExt, ID3D12GraphicsCommandListExt

### PR DESCRIPTION
**Pull request by rgc183**
These interfaces are needed for d3d12 NvAPI support. Adds tests for new APIs.

Signed-off-by: Roshan Chaudhari <rochaudhari@nvidia.com>

Reviewed-by: Liam Middlebrook <lmiddlebrook@nvidia.com>

### CODE REVIEW

**Comment by liam-middlebrook** [comment]
I think it may make more sense for the test additions to be in a separate commit from the new interfaces.

**Comment by liam-middlebrook** [comment]
@jp7677 you may want to take a look at this also, even if just for a heads up that we're trying to go for roughly the same structure as DXVK's extension mechanism that dxvk-nvapi uses for the `NvAPI_D3D*` functions.

**Comment by rgc183** [comment]
The reason for single commit for the feature and tests are we can revert it in single commit. I think it is also easier to review the tests since we do not need to refer the feature again in case of different commit. However, I am okay to make it separate if everyone votes for it.

File: include/vkd3d_device_ext.idl
Comment (HansKristian-Work) [inline include/vkd3d_device_ext.idl:24]: Weird newline

File: include/vkd3d_device_ext.idl
Comment (HansKristian-Work) [inline include/vkd3d_device_ext.idl:35]: Pointer asterisk binds right

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:8815]: Pointer binds right

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:8821]: Seems a bit weird to chain these. Would be better I think to call d3d12_command_list_AddRef directly.

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:8828]: d3d12_command_list_Release

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:8839]: Not sure how COM rules work here, but IIRC we have to be able to query ID3D12GraphicsCommandList interfaces here.

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:8854]: d3d12_command_list_GetDevice

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:8866]: Weird newline

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [reply libs/vkd3d/command.c:8854]: Also not sure why this is useful. You could just have called GetDevice directly from ID3D12GraphicsCommandList instead of chaining via the Ext interface

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:8869]: Seems a bit overkill to make this a function that can fail. Just assign the vtable pointer directly along with the normal interface.

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:8895]: Space between `if` and `(`, but this code should be nuked either way.

File: libs/vkd3d/device.c
Comment (HansKristian-Work) [inline libs/vkd3d/device.c:2381]: Space

File: libs/vkd3d/device.c
Comment (HansKristian-Work) [inline libs/vkd3d/device.c:5000]: Pointer

File: libs/vkd3d/device.c
Comment (HansKristian-Work) [inline libs/vkd3d/device.c:5021]: Again, I think we need to be able to query original interfaces from this interface. Perhaps we could just chain QueryInterface to the ID3D12Device impl

File: libs/vkd3d/device.c
Comment (HansKristian-Work) [inline libs/vkd3d/device.c:5035]: pointer

File: libs/vkd3d/device.c
Comment (HansKristian-Work) [inline libs/vkd3d/device.c:5045]: `/* */` comments only

File: libs/vkd3d/vkd3d_private.h
Comment (HansKristian-Work) [inline libs/vkd3d/vkd3d_private.h:1742]: Weird spaces

File: tests/d3d12_ext_test.c
Comment (HansKristian-Work) [inline tests/d3d12_ext_test.c:27]: pointer

File: tests/d3d12_ext_test.c
Comment (HansKristian-Work) [inline tests/d3d12_ext_test.c:38]: missing spaces between ,

File: tests/d3d12_ext_test.c
Comment (HansKristian-Work) [reply tests/d3d12_ext_test.c:27]: Variables should be sorted in inverse christmas tree order

File: tests/d3d12_ext_test.c
Comment (HansKristian-Work) [inline tests/d3d12_ext_test.c:114]: C89 comments

File: tests/d3d12_ext_test.c
Comment (HansKristian-Work) [inline tests/d3d12_ext_test.c:51]: C89 comments

File: include/vkd3d_device_ext.idl
Comment (HansKristian-Work) [reply include/vkd3d_device_ext.idl:35]: Don't you need VkDevice as well?

File: include/vkd3d_command_list_ext.idl
Comment (HansKristian-Work) [inline include/vkd3d_command_list_ext.idl:29]: Not sure why you need GetDevice here since ID3D12GraphicsCommandList already has that, and surprised there is no GetVulkanHandle for VkCommandBuffer.

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
Mostly code nits. My COM-fu is not very strong, so @Joshua-Ashton and @doitsujin should have a look.

**Comment by HansKristian-Work** [comment]
I don't particularly mind having the test in the same commit.

File: include/vkd3d_device_ext.idl
Comment (Joshua-Ashton) [inline include/vkd3d_device_ext.idl:37]: Newline at end of file, ditto for all the others.

File: libs/vkd3d/command.c
Comment (Joshua-Ashton) [reply libs/vkd3d/command.c:8839]: Yes, we need to return the parent here too under its guids.

File: libs/vkd3d/device.c
Comment (Joshua-Ashton) [inline libs/vkd3d/device.c:5007]: d3d12_device_AddRef

File: include/vkd3d_device_ext.idl
Comment (Joshua-Ashton) [reply include/vkd3d_device_ext.idl:34]: Please no `void *`. Just forward declare the handles for the idl and cpp_quote including vulkan/vulkan_core.h

File: libs/vkd3d/device.c
Comment (Joshua-Ashton) [inline libs/vkd3d/device.c:5062]: Please can we be consistent with our naming here?

I'd also prefer it if we called stuff something like
`d3d12_device_vkd3d_ext` or something (same with the command list), just for clarity when reading that this isn't a part of D3D12.

File: libs/vkd3d/device.c
Comment (Joshua-Ashton) [inline libs/vkd3d/device.c:5038]: Make this function type safe

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: include/vkd3d_command_list_vkd3d_ext.idl
Comment (HansKristian-Work) [inline include/vkd3d_command_list_vkd3d_ext.idl:19]: @Joshua-Ashton Do we need cpp_quote here if we're forward declaring the Vulkan types anyways?

File: libs/vkd3d/vkd3d_private.h
Comment (HansKristian-Work) [inline libs/vkd3d/vkd3d_private.h:1742]: Weird space

File: libs/vkd3d/device.c
Comment (HansKristian-Work) [inline libs/vkd3d/device.c:5045]: Two spaces between `D3D12_VK_EXTENSION` and `extension`

File: libs/vkd3d/device.c
Comment (HansKristian-Work) [inline libs/vkd3d/device.c:5036]: Missing space

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:8852]: Missing space

File: include/vkd3d_device_vkd3d_ext.idl
Comment (HansKristian-Work) [inline include/vkd3d_device_vkd3d_ext.idl:27]: Shouldn't the name of this enum value be something D3D12_VK_* related?

File: include/vkd3d_command_list_vkd3d_ext.idl
Comment (rgc183) [reply include/vkd3d_command_list_vkd3d_ext.idl:19]: You are right. Not needed. Will remove.

File: libs/vkd3d/command.c
Comment (HansKristian-Work) [inline libs/vkd3d/command.c:8890]: It feels somewhat unnecessary to have a separate struct here which requires chaining two CONTAINING_RECORDs. `list->ID3D12GraphicsCommandListExt_iface` seems sensible to me.

File: libs/vkd3d/vkd3d_private.h
Comment (rgc183) [reply libs/vkd3d/vkd3d_private.h:1742]: this is weird :) actually there is a space in original file after `VkDeviceAddress predicate_va;`

File: libs/vkd3d/device.c
Comment (HansKristian-Work) [inline libs/vkd3d/device.c:5025]: This violates the COM rule of:

> For any given object instance, a call to QueryInterface with IID_IUnknown must always return the same physical pointer value.

Wouldn't it work to just chain directly to d3d12_device_QueryInterface?

File: include/vkd3d_device_vkd3d_ext.idl
Comment (rgc183) [reply include/vkd3d_device_vkd3d_ext.idl:27]: This is just dummy enum. I should rename it for time being.

File: libs/vkd3d/device.c
Comment (rgc183) [reply libs/vkd3d/device.c:5025]: Good catch.!

File: meson.build
Comment (HansKristian-Work) [inline meson.build:78]: I assume this is not needed.

Comment (rgc183) [reply meson.build:78]: Yes, once I remove `include` from `.idl`.

File: libs/vkd3d/command.c
Comment (rgc183) [reply libs/vkd3d/command.c:8890]: hmm, yeah looks like currently we do not have any other member to add, makes sense to remove redundant chaining. Will do.

Comment (rgc183) [reply libs/vkd3d/command.c:8890]: same goes for the device.

File: libs/vkd3d/command.c
Comment (doitsujin) [reply libs/vkd3d/command.c]: Yeah, the proper way to do this is to implement all COM queries in one place (`d3d12_command_list`) and have this function call that instead.

File: include/vkd3d_command_list_vkd3d_ext.idl
Comment (HansKristian-Work) [inline include/vkd3d_command_list_vkd3d_ext.idl:28]: Do we know that this is enough? Is the plan to add CommandListExt1 etc later if we realize something is missing?

**Review by HansKristian-Work — APPROVED** [review]
LGTM. I'll consider it approved when @Joshua-Ashton or @doitsujin approve as well.

File: include/vkd3d_command_list_vkd3d_ext.idl
Comment (rgc183) [reply include/vkd3d_command_list_vkd3d_ext.idl:28]: If you mean methods, surely we need to more here as needed. At this point we do not have list to be added.

Comment (HansKristian-Work) [reply include/vkd3d_command_list_vkd3d_ext.idl:28]: Right, but that means ABI breakage? I assume if more interfaces are required, we'd have to add new derived interfaces in the future.

**Comment by HansKristian-Work** [comment]
Please squash the patch set beforehand.

File: include/vkd3d_command_list_vkd3d_ext.idl
Comment (Joshua-Ashton) [reply include/vkd3d_command_list_vkd3d_ext.idl]: No

**Comment by Joshua-Ashton** [comment]
Can you please add what you need for DLSS into the base interface so we don't do needless bumps before merging?

**Comment by rgc183** [comment]
@HansKristian-Work @Joshua-Ashton Makes sense. I will update this PR with what DLSS needs.

**Comment by rgc183** [comment]
Just a note, update to this PR going to take a while once we get internal sign offs.

File: tests/d3d12.c
Comment (Joshua-Ashton) [inline tests/d3d12.c:112]: Please make moving this in a separate commit.

File: tests/d3d12_ext_test.c
Comment (Joshua-Ashton) [inline tests/d3d12_ext_test.c]: Newline at end of file.

**Comment by HansKristian-Work** [comment]
Obsoleted by #770, closing.

**Pull request closed without merging**
